integration/app_integration_test.go

```diff
@@ -0,0 +1,430 @@
+	})
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	req.Header.Set(csrf.HeaderName, csrfToken)
+
+	// Issue request.
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+	resp, err := client.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	// Read in response.
+	var csResp *web.CreateSessionResponse
+	err = json.NewDecoder(resp.Body).Decode(&csResp)
+	require.NoError(t, err)
+
+	// Extract session cookie and bearer token.
+	require.Len(t, resp.Cookies(), 1)
+	cookie := resp.Cookies()[0]
+	require.Equal(t, cookie.Name, web.CookieName)
+
+	return cookie.Value, csResp.Token
+}
+
+// createAppSession creates an application session with the root cluster. The
+// application that the user connects to may be running in a leaf cluster.
+func (p pack) createAppSession(t *testing.T, sessionCookie string, bearerToken string, publicAddr, clusterName string) string {
+	casReq, err := json.Marshal(web.CreateAppSessionRequest{
+		FQDN:        publicAddr,
+		PublicAddr:  publicAddr,
+		ClusterName: clusterName,
+	})
+	require.NoError(t, err)
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(Loopback, p.rootCluster.GetPortWeb()),
+		Path:   "/v1/webapi/sessions/app",
+	}
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewBuffer(casReq))
+	require.NoError(t, err)
+
+	req.AddCookie(&http.Cookie{
+		Name:  web.CookieName,
+		Value: sessionCookie,
+	})
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %v", bearerToken))
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+	resp, err := client.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	var casResp *web.CreateAppSessionResponse
+	err = json.NewDecoder(resp.Body).Decode(&casResp)
+	require.NoError(t, err)
+
+	return casResp.CookieValue
+}
+
+// makeRequest makes a request to the root cluster with the given session cookie.
+func (p *pack) makeRequest(sessionCookie string) (int, string, error) {
+	u := url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(Loopback, p.rootCluster.GetPortWeb()),
+		Path:   "/",
+	}
+	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+	if err != nil {
+		return 0, "", trace.Wrap(err)
+	}
+	req.AddCookie(&http.Cookie{
+		Name:  app.CookieName,
+		Value: sessionCookie,
+	})
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return 0, "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	// Read in response body and make sure it contains the expected message.
```

Comment:
nit: the following does not actually make sure the response contains the expected message.